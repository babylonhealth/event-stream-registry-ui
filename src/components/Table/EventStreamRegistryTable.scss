@import 'material-icons/iconfont/material-icons.scss';
@import "../EventStreamRegistry/_variables.scss";

$textFontSize: 10px;
$iconFontSize: 10px;
$tootipFontSize: 12px;

$fontFamily: "Courier New", sans-serif;

.material-icons.icon {
  font-size: $iconFontSize;
}

.event-registry-table {
  width: 100vw;
  height: calc(100% - 25px); /* - banner height */
  overflow: scroll;
  position: fixed;
}

table {
  display: -moz-groupbox;
  table-layout: fixed;
  border-collapse: collapse;
  border-spacing: 0;
}

th {
  text-align: left;
  font-size: $textFontSize;
  font-family: $fontFamily;
}

tr {
  border-spacing: 0;
  width: 100%;
  display: inline-table;
  table-layout: fixed;
}

tr:hover {
  font-weight: bold;
}

td {
  line-height: 0;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  cursor: default;
  font-family: $fontFamily;
  font-size: $textFontSize;
}

.id {
  width: 300px;
  left: 0;
  position: sticky;
  padding-left: $lpad;
  background-color: white;
  z-index: 100;
}

.status {
  width: 15px;
  font-size: $textFontSize;
  padding-right: 5px;
}

.url {
  width: 15px;
  text-align: center;
  padding-right: 5px;
}
.url a {
  color: $urlBlue;
}
.url a:visited {
  color: $urlBlue;
}

.metadata {
  width: 150px;
}

.metadata a {
  text-decoration: none;
  color: $urlBlue;
}

.metadata a:visited {
  color: $urlBlue;
}

.tooltip {
  visibility: hidden;
  opacity: 0;
  position: absolute;
  display: block;
  margin-top: 15px;
  margin-left: 20px;
  line-height: 20px;
  z-index: 100;
  pointer-events: none;
  padding-left: 5px;
  padding-right: 5px;
  background-color: white;
  color: black;
  border: 0.1px solid $lightGrey;
  font-family: $fontFamily;
  font-size: $tootipFontSize;
}

td:hover .tooltip {
  visibility: visible;
  opacity: 1;
  -webkit-transition-delay: 2s;
  -moz-transition-delay: 2s;
  -o-transition-delay: 2s;
  transition: opacity .5s;
  transition-delay: .5s;
}
