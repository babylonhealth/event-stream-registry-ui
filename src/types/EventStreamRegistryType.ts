/**
 * The Event Stream Registry dataset.
 *
 * @param node - contains the set of services and topics in the Event Stream Registry.
 * @param links - defines the connectedness of the nodes.
 * @type {EventStreamRegistryType}
 */
export interface EventStreamRegistryType {
    nodes: EventStreamRegistryNode[],
    links: EventStreamRegistryLink[]
}

/**
 * A single Event Stream Registry node - either a topic or a service.
 *
 * @param id - the node id
 * @param type - the node type, either "SERVICE" or "TOPIC"
 * @param metadata - a map node metadata, the key is the metadata name, the value is a Metadata object
 * @param urls - a list of associated node urls
 * @type {EventStreamRegistryNode}
 * */
export interface EventStreamRegistryNode {
    id: string,
    type: NodeType,
    metadata: Record<string, Metadata>,
    urls: Url[],
}

/**
 * A metadata value.
 *
 * @param value - the metadata value
 * @param type - either 'NUMBER' or 'STRING', the metadata type, the type defines allowed filtering operations
 * @param url? - an associated url, renders values as links
 * @param status? - either 'GOOD' or 'BAD', indicates the "health" of a metric, e.g. a high consumer lag might be
 * consider 'unhealthy` of `BAD`, renders values `green` and `red`.
 * @type {Metadata}
 * */
export interface Metadata {
    "value": string | number,
    "type": MetadataType
    "url": string | undefined
    "status": Status | undefined
}

/**
 * An associated service or topic url.
 *
 * @param name - the name of the url, e.g. grafana
 * @param value - the url, e.g. https://my.grafana.co.uk/event-stream-dashboard/my-topic
 * @param icon? - a material ui icon to render urls, if null urls are rendered as 'launch' icon
 * @type {Url}
 * */
export interface Url {
    "name": string,
    "value": string,
    "icon": string | undefined
}

/**
 * Defines a link between two nodes
 *
 * @param source - the node id of the source
 * @param target - the node id of the target
 * @type {EventStreamRegistryLink}
 * */
export interface EventStreamRegistryLink {
    source: string,
    target: string,
}

export enum MetadataType {
    "NUMBER" = "NUMBER",
    "STRING" = "STRING"
}

export enum Status {
    "BAD",
    "GOOD",
}

export enum StatusColors {
    BAD = "#FF0000",
    GOOD = "#008000",
    NO_STATUS = "#000000",
}

export enum NodeType {
    "TOPIC" = "TOPIC",
    "SERVICE" = "SERVICE"
}
