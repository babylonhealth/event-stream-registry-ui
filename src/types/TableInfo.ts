export type MetadataFields = {
    numeric: Set<string>,
    string: Set<string>,
    url: Set<string>
}

export type EventStreamRegistryMetadata = {
    topic: MetadataFields,
    service: MetadataFields
}
export type QueryParameter = {
    key: string,
    value: string
    operation: string
}
export enum View {
    "GRAPH"= "GRAPH",
    "SERVICES"= "SERVICES",
    "TOPICS"= "TOPICS",
    "HELP"= "HELP"
}

