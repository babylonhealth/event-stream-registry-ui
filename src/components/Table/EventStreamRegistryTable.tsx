import * as React from "react";

import {IdHeader, MetadataHeader, StatusHeader, UrlHeader, IdCell, MetadataCell, StatusCell, UrlCell} from "./EventStreamRegistryTableCell";

import {MetadataFields} from "../../types/TableInfo";
import {EventStreamRegistryNode} from "../../types/EventStreamRegistryType";

import "./EventStreamRegistryTable.scss"


type EventStreamRegistryTableProps = {
    nodes: EventStreamRegistryNode[]
    tableInfo: MetadataFields
}

/**
 * Renders a html table of node information.
 * The EventStreamRegistry table implements either TOPIC nodes or SERVICE nodes.
 *
 * @param {EventStreamRegistryNode[]} nodes - a list of nodes
 * @param {MetadataFields} tableInfo - the set of numeric, string, and url of either SERVICE or TOPIC metadata fields
 */
const EventStreamRegistryTable = ({nodes, tableInfo}: EventStreamRegistryTableProps) => {

    const numericFields: string[] = Array.from(tableInfo.numeric);
    const stringFields: string[] = Array.from(tableInfo.string);
    const urlFields: string[] = Array.from(tableInfo.url);

    return (
        <div className="event-registry-table">
            <table>
                <thead>
                <tr>
                    <IdHeader/>
                    <StatusHeader/>
                    {urlFields.map((field, i) => <UrlHeader key={i}/>)}
                    {stringFields.map((field, i) => <MetadataHeader key={i} name={field}/>)}
                    {numericFields.map((field, i) => <MetadataHeader key={i} name={field}/>)}
                </tr>
                </thead>
                <tbody>
                {nodes
                    .sort((a, b) => (a.id > b.id) ? 1 : ((b.id > a.id) ? -1 : 0))
                    .map((node, i) =>
                        <tr key={i}>
                            <IdCell id={node.id}/>
                            <StatusCell node={node}/>
                            {urlFields.map((field, j) =>
                                <UrlCell key={j} url={node.urls.find(l => l.name === field)}/>)}
                            {stringFields.map((field, j) =>
                                <MetadataCell key={j} metadata={node.metadata[field]}/>)}
                            {numericFields.map((field, j) =>
                                <MetadataCell key={j} metadata={node.metadata[field]}/>)}
                        </tr>
                    )}
                </tbody>
            </table>
        </div>
    );
};

export default EventStreamRegistryTable;
