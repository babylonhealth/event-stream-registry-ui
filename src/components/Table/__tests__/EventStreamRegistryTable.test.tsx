import React from 'react'
import {cleanup, render} from '@testing-library/react'
import {MetadataFields} from "../../../types/TableInfo";
import EventStreamRegistryTable from "../EventStreamRegistryTable";
import {EventStreamRegistryNode, MetadataType, NodeType, Status} from "../../../types/EventStreamRegistryType";

afterEach(cleanup);

it('should take a snapshot', () => {

    const nodes: EventStreamRegistryNode[] = [
            {
                'id': 'topic1',
                'type': NodeType.TOPIC,
                'metadata': {
                    "numeric_metadata_field_with_status_and_url": {
                        "value": 0,
                        "type": MetadataType.NUMBER,
                        "url": "www.example.co.uk",
                        "status": Status.GOOD
                    },
                    "numeric_metadata_field_with_status": {
                        "value": 1,
                        "type": MetadataType.NUMBER,
                        "url": undefined,
                        "status": Status.BAD
                    },
                    "numeric_metadata_field_with_url": {
                        "value": 2,
                        "type": MetadataType.NUMBER,
                        "url": "www.example.co.uk",
                        "status": undefined
                    },
                    "string_metadata_field_with_status_and_url": {
                        "value": "0",
                        "type": MetadataType.STRING,
                        "url": "www.example.co.uk",
                        "status": Status.GOOD
                    },
                    "string_metadata_field_with_status": {
                        "value": "1",
                        "type": MetadataType.STRING,
                        "url": undefined,
                        "status": Status.BAD
                    },
                    "string_metadata_field_with_url": {
                        "value": "2",
                        "type": MetadataType.STRING,
                        "url": "www.example.co.uk",
                        "status": undefined
                    },
                },
                'urls': [
                    {
                        "name": "url_with_icon",
                        "value": "www.bablontech.co.uk",
                        "icon": "special_icon",
                    },
                    {
                        "name": "url_without_icon",
                        "value": "www.bablontech.co.uk",
                        "icon": undefined,
                    },
                ]
            },
            {
                'id': 'topic2',
                'type': NodeType.TOPIC,
                'metadata': {
                    "numeric_metadata_field_with_status_and_url": {
                        "value": 0,
                        "type": MetadataType.NUMBER,
                        "url": "www.example.co.uk",
                        "status": Status.GOOD
                    },
                    "numeric_metadata_field_with_status": {
                        "value": 1,
                        "type": MetadataType.NUMBER,
                        "url": undefined,
                        "status": Status.BAD
                    },
                    "numeric_metadata_field_with_url": {
                        "value": 2,
                        "type": MetadataType.NUMBER,
                        "url": "www.example.co.uk",
                        "status": undefined
                    }
                },
                'urls': []
            }
        ]
    ;

    const tableInfo: MetadataFields = {
        'numeric': new Set(["numeric_metadata_field"]),
        'string': new Set(["string_metadata_field"]),
        'url': new Set(),
    };

    const {asFragment} = render(<EventStreamRegistryTable nodes={nodes} tableInfo={tableInfo}/>);
    // @ts-ignore
    expect(asFragment(<EventStreamRegistryTable nodes={nodes} tableInfo={tableInfo}/>)).toMatchSnapshot()
});
