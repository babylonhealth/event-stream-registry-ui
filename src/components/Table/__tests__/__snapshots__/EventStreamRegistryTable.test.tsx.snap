// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="event-registry-table"
  >
    <table>
      <thead>
        <tr>
          <th
            class="id"
            scope="col"
          />
          <th
            class="status"
            scope="col"
          />
          <th
            class="metadata"
            scope="col"
          >
            string metadata field
          </th>
          <th
            class="metadata"
            scope="col"
          >
            numeric metadata field
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="id"
          >
            topic1
          </td>
          <td
            class="status"
          >
            <span
              class="material-icons icon"
              style="color: rgb(255, 0, 0);"
            >
              circle
            </span>
          </td>
          <td
            class="metadata"
            style="color: rgb(0, 0, 0);"
          >
            <span
              class="tooltip"
            />
          </td>
          <td
            class="metadata"
            style="color: rgb(0, 0, 0);"
          >
            <span
              class="tooltip"
            />
          </td>
        </tr>
        <tr>
          <td
            class="id"
          >
            topic2
          </td>
          <td
            class="status"
          >
            <span
              class="material-icons icon"
              style="color: rgb(255, 0, 0);"
            >
              circle
            </span>
          </td>
          <td
            class="metadata"
            style="color: rgb(0, 0, 0);"
          >
            <span
              class="tooltip"
            />
          </td>
          <td
            class="metadata"
            style="color: rgb(0, 0, 0);"
          >
            <span
              class="tooltip"
            />
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
